bothan-binance/src/worker.rs

```diff
@@ -0,0 +1,78 @@
+use tokio::sync::mpsc::Sender;
+
+use bothan_core::store::Store;
+use bothan_core::worker::{AssetStatus, AssetWorker, Error};
+
+use crate::api::websocket::BinanceWebSocketConnector;
+
+mod asset_worker;
+pub mod builder;
+pub(crate) mod error;
+mod types;
+
+/// A worker that fetches and stores the asset information from Binance's API.
+pub struct BinanceWorker {
+    connector: BinanceWebSocketConnector,
+    store: Store,
+    subscribe_tx: Sender<Vec<String>>,
+    unsubscribe_tx: Sender<Vec<String>>,
+}
+
+impl BinanceWorker {
+    /// Create a new worker with the specified connector, store and channels.
+    pub fn new(
+        connector: BinanceWebSocketConnector,
+        store: Store,
+        subscribe_tx: Sender<Vec<String>>,
+        unsubscribe_tx: Sender<Vec<String>>,
+    ) -> Self {
+        Self {
+            connector,
+            store,
+            subscribe_tx,
+            unsubscribe_tx,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl AssetWorker for BinanceWorker {
+    /// Fetches the AssetStatus for the given cryptocurrency ids.
+    async fn get_assets<T: AsRef<str> + Send + Sync>(&self, ids: &[T]) -> Vec<AssetStatus> {
+        self.store.get_assets(ids).await
+    }
+
+    /// Adds the specified cryptocurrency IDs to the query set and subscribes to their updates.
+    async fn add_query_ids<T: AsRef<str> + Send + Sync>(&self, ids: &[T]) -> Result<(), Error> {
+        let to_sub = self
+            .store
+            .filter_existing_query_ids(ids.iter().map(|id| id.as_ref()).collect::<Vec<&str>>())
```

Comment:
I think the api of `store` should have `add_query_ids` that adds and returns the ids that are added so it's atomic + `remove_query_ids` that returns the ids that are removed.

We don't really expect `self.subscribe_tx.send` ever so i think it's ok to mutate the store before sending the sub message.

Reply:
modified, but will retroactively remove from store on failure